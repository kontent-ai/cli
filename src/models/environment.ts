export interface IEnvironment {
    readonly environmentId: string;
    readonly apiKey: string;
}

export interface IEnvironmentsConfig {
    [index: string]: IEnvironment;
}
