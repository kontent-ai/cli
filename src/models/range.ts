export interface IRange {
    from: number;
    to: number;
}
