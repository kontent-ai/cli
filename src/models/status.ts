export interface IMigrationStatus {
    readonly name: string;
    readonly success: boolean;
    readonly order: number | Date;
    readonly time: Date;
    readonly lastOperation?: Operation;
}

export interface IStatus {
    [environmentId: string]: IMigrationStatus[];
}

export type Operation = 'run' | 'rollback';
