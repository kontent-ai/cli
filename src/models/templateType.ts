export enum TemplateType {
    TypeScript,
    Javascript,
}
