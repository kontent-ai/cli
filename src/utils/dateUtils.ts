export const formatDateForFileName = (date: Date) =>
    `${date.getUTCFullYear()}-` +
    `${('0' + (date.getUTCMonth() + 1)).slice(-2)}-` +
    `${('0' + date.getUTCDate()).slice(-2)}-` +
    `${('0' + date.getUTCHours()).slice(-2)}-` +
    `${('0' + date.getUTCMinutes()).slice(-2)}-` +
    `${('0' + date.getUTCSeconds()).slice(-2)}-`;
