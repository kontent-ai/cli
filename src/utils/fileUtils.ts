import fs, { PathLike } from 'fs';
import * as path from 'path';

export const fileExists = (filePath: PathLike): boolean => {
    return fs.existsSync(filePath);
};

export const isAllowedExtension = (filename: string): boolean => {
    const extension = path.extname(filename);
    return ['.js', ''].includes(extension);
};

export const getFileWithExtension = (filename: string, defaultExtension: string = '.js'): string => {
    const hasFileExtension = path.extname(filename);

    if (hasFileExtension) {
        const normalized = filename.split(defaultExtension).slice(0, -1).join(defaultExtension);
        return normalized + defaultExtension;
    } else {
        return filename + defaultExtension;
    }
};

export const getFileBackupName = () => {
    const currentDate = new Date();
    const formatted = `${currentDate.getFullYear()}-${currentDate.getMonth()}-${currentDate.getDate()}-${currentDate.getTime()}`;
    return `backup-${formatted}`;
};
