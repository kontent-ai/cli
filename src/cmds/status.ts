import yargs from 'yargs';
import { createStatusImplementationFile } from '../utils/statusManager';

const statusCommand: yargs.CommandModule = {
    command: 'status',
    describe: 'Status commands',
    builder: (yargs: any) =>
        yargs
            .options({
                implementation: {
                    alias: 'i',
                    describe: 'Implement your own save/read functions to deal with status.json.',
                    type: 'boolean',
                },
            })
            .demandOption(['implementation']),
    handler: (argv: any) => {
        if (argv.implementation) {
            createStatusImplementationFile();
        }
    },
};

Object.assign(exports, statusCommand);
