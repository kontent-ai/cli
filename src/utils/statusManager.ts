import { IMigrationStatus, IStatus } from '../models/status';
import { readFileSync, writeFileSync, mkdirSync } from 'fs';
import { fileExists } from './fileUtils';
import * as path from 'path';
import { loadStatusPlugin } from './status/statusPlugin';

const migrationStatusFilename = 'status.json';
const statusDirectoryName = 'status';
const statusImplementationFilename = 'statusImpl.ts';
let status: IStatus = {};

const updateMigrationStatus = async (projectId: string, migrationStatus: IMigrationStatus) => {
    let projectMigrationsHistory = status[projectId];

    if (projectMigrationsHistory === undefined) {
        projectMigrationsHistory = status[projectId] = [];
    }

    const previousMigrationStatusIndex = projectMigrationsHistory.findIndex((executedMigration) => executedMigration.name === migrationStatus.name);
    if (previousMigrationStatusIndex > -1) {
        projectMigrationsHistory[previousMigrationStatusIndex] = migrationStatus;
    } else {
        projectMigrationsHistory.push(migrationStatus);
    }

    await saveStatusFile();
};

export const markAsCompleted = async (projectId: string, name: string, order: number | Date) => {
    const migrationStatus = {
        name,
        order,
        success: true,
        time: new Date(Date.now()),
    };

    await updateMigrationStatus(projectId, migrationStatus);
};

const saveStatusFile = async () => {
    const statusJSON = JSON.stringify(status, null, 2);

    if (fileExists(getStatusImplementationFilePath())) {
        try {
            const file = await loadStatusPlugin(getStatusImplementationFilePath().slice(0, -3) + '.js');
            await file.saveStatus(statusJSON);
        } catch (e) {
            console.log(`Could not load the plugin due to ${e}. Fallbacking to status.json`);
        } finally {
            return;
        }
    }
    saveStatusToFile(statusJSON);
};

const getMigrationStatus = (migrationName: string, projectId: string): IMigrationStatus | undefined => {
    const projectStatus = status[projectId];
    if (projectStatus == null) {
        return undefined;
    }

    return projectStatus.find((migrationStatus) => migrationStatus.name === migrationName);
};

export const wasSuccessfullyExecuted = (migrationName: string, projectId: string): boolean => {
    const result = getMigrationStatus(migrationName, projectId);
    return !!result && result.success;
};

const getStatusFilepath = (): string => {
    return path.join(process.cwd(), migrationStatusFilename);
};

export const loadMigrationsExecutionStatus = async () => {
    if (fileExists(getStatusImplementationFilePath())) {
        try {
            const file = await loadStatusPlugin(getStatusImplementationFilePath().slice(0, -3) + '.js');
            status = await file.readStatus();
        } catch (e) {
            console.log(`Could not load the plugin due to ${e}`);
        } finally {
            return;
        }
    }
    status = readFromStatus();
};

const readFromStatus = (): IStatus => {
    const statusFilepath = getStatusFilepath();
    if (!fileExists(statusFilepath)) {
        return {};
    }

    try {
        const projectsMigrationStatuses = readFileSync(getStatusFilepath()).toString();

        return JSON.parse(projectsMigrationStatuses);
    } catch (error) {
        console.warn(`Status JSON file is invalid because of ${error instanceof Error ? error.message : 'unknown error.'}. Continuing with empty status.`);
        return {};
    }
};

const saveStatusToFile = (data: string): void => {
    const statusFilepath = getStatusFilepath();

    try {
        writeFileSync(statusFilepath, data, { flag: 'w' });
        console.log(`Status file was updated see ${statusFilepath}`);
    } catch (error) {
        console.error(`Status file save failed, because of ${error instanceof Error ? error.message : 'unknown error.'}`);
    }
};

const getStatusImplementationFilePath = () => path.join(process.cwd(), statusDirectoryName, statusImplementationFilename);

const getStatusImplementationDirectoryPath = () => path.join(process.cwd(), statusDirectoryName);

export const createStatusImplementationFile = () => {
    const statusImplementationPath = getStatusImplementationFilePath();

    if (fileExists(statusImplementationPath)) {
        console.error(`File ${statusImplementationPath} already exists`);
    }

    try {
        if (!fileExists(getStatusImplementationDirectoryPath())) {
            mkdirSync(getStatusImplementationDirectoryPath());
        }

        writeFileSync(statusImplementationPath, statusImplementationTemplate, { flag: 'w' });
        console.log(`Status.ts file was created see ${statusImplementationPath}`);
    } catch (error) {
        console.error(`Status.ts file creation failed, because of ${error instanceof Error ? error.message : 'unknown error.'}`);
    }
};

const statusImplementationTemplate = `
import type {IStatus} from "@kontent-ai/cli";

export const saveStatus = async (data: string) => {

}

export const readStatus = async (): IStatus => {

}
`;
