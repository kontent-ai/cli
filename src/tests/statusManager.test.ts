import * as statusManager from '../utils/statusManager';
import * as fileUtils from '../utils/fileUtils';
import * as fs from 'fs';
import * as path from 'path';
import { IStatus } from '../models/status';

const readStatusFile = (): IStatus => {
    const statusFilepath = path.join(process.cwd(), 'status.json');

    const fileContent = fs.readFileSync(statusFilepath).toString();
    return JSON.parse(fileContent);
};

describe('Status manager', () => {
    beforeAll(() => {
        jest.spyOn(Date, 'now').mockImplementation(() => 1575939660000);
    });

    it('Project has success status in status manager file', async () => {
        const projectId = 'project1';
        const migrationName = 'migration1';
        await statusManager.markAsCompleted({}, projectId, migrationName, 1, 'run', null);

        const statusFile = readStatusFile();
        const status = statusFile[projectId][0];

        expect(status).toMatchSnapshot();
    });

    it('Not executed migration is not present in status file', async () => {
        const project1Id = 'project1';
        const project2Id = 'project2';
        const migration1Name = 'migration1';
        await statusManager.markAsCompleted({}, project1Id, migration1Name, 1, 'run', null);

        const projectMigrationStatus = statusManager.shouldSkipMigration({}, migration1Name, project2Id, 'run');

        expect(projectMigrationStatus).toBe(false);
    });

    it('Executed migration is present in status file', async () => {
        const project2Id = 'project2';
        const migration2Name = 'migration2';
        await statusManager.markAsCompleted({}, project2Id, migration2Name, 1, 'run', null);

        const projectMigrationStatus = statusManager.shouldSkipMigration({}, project2Id, migration2Name, 'run');

        expect(projectMigrationStatus).toBe(false);
    });

    it('loadMigrationsExecutionStatus to be called with plugins', async () => {
        const expectedStatus = { '30816c62-8d41-4dc4-a1ab-40440070b7bf': [] };

        const readStatusPlugin = async () => {
            return expectedStatus;
        };

        const returnedStatus = await statusManager.loadMigrationsExecutionStatus(readStatusPlugin);

        expect(returnedStatus).toEqual(expectedStatus);
    });

    it('loadMigrationsExecutionStatus to be called with plugin that throw', async () => {
        const readStatusPlugin = async () => {
            throw new Error('Error during plugin function');
        };

        expect.assertions(1);

        return statusManager.loadMigrationsExecutionStatus(readStatusPlugin).catch((e) => expect((e as Error).message).toMatch('Error during plugin function'));
    });

    it('MarkAsCompleted to be called with plugins', async () => {
        jest.spyOn(fileUtils, 'fileExists').mockReturnValue(true);

        const saveStatusMocked = jest.fn().mockImplementation(() => Promise.resolve());

        await statusManager.markAsCompleted({}, '', 'testMigration', 1, 'run', saveStatusMocked);

        expect(saveStatusMocked).toHaveBeenCalled();
    });
});
