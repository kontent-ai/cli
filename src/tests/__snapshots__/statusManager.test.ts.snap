// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status manager Project has success status in status manager file 1`] = `
Object {
  "lastOperation": "run",
  "name": "migration1",
  "order": 1,
  "success": true,
  "time": "2019-12-10T01:01:00.000Z",
}
`;
